Add code review guidelines (#17)
bitwarden/contributing-docs: docs/contributing/pull-requests/feature-branches.md

> # Long-lived Feature Branches

Reviewer: Will this page ever be relevant to community contributors? What do you thinking about making it Bitwarden-only?

> A long lived feature branch differ from regular Pull Requests in three significant ways.

Reviewer: ```suggestion
A long lived feature branch differs from regular Pull Requests in three significant ways:
```

> 1. Has multiple authors.
> 2. Consists of multiple already reviewed Pull Requests.
> 3. Only consists of merge commits.

Reviewer: ```suggestion
1. It has multiple authors.
2. It consists of multiple already reviewed Pull Requests.
3. It only consists of merge commits.
```

Author: What do you think about changing the the leading sentance and the lists to:

```
A long lived feature branch differs from a regular Pull Requests in three significant ways. It
should:

- Have multiple authors.
- Consists of multiple already reviewed Pull Requests.
- Only consists of merge commits.
```

I also changed the list from numbered since the order doesn't matter.

Reviewer: Yes, also a good solution. I'd also drop the full-stops at the end of each dot point but that's up to you.

Reviewer: I've left a new review below

    testing.
- Review any non-reviewed commits made directly on the branch, these can be either feature work or
  merge commits. For more details see [Reviewing](#reviewing).

:::warning

Since feature branches do not have the same protections as master, it's technically possible to
commit directly to the branch or merge a pull request without a up-to-date review. However this
should be discouraged, and should be avoided whenever possible. The only exception being merge
commits.

:::

## Syncing Feature Branch

We typically denote one person as being responsible for keeping the feature branch up to
date.<bitwarden> In most cases this will be the TPC of the pod but can be anyone.</bitwarden> This
person is responsible for keeping the feature branch reasonably up to date with master, preparing it
for merge and more.

Due to how GitHub handles reviews, this person cannot also approve the feature branch. The reviewer
of the feature branch can be anyone in the pod. However, since the work in a feature branch is
typically larger, some seniority with the codebase can be beneficial.

## Reviewing

We can perform the review using GitHub's UI. By opening the Pull Request and clicking on the

> `Commits` tab. Afterwards check each individually commit.

Reviewer 2: ```suggestion
`Commits` tab. Afterwards check each commit individually.
```

---
sidebar_custom_props:
  access: bitwarden
---

# Long-lived Feature Branches

> A long lived feature branch differs from regular Pull Requests in three significant ways.

Reviewer 3: ```suggestion
A long-lived feature branch differs from regular Pull Requests in three significant ways.
```

1. Has multiple authors.
2. Consists of multiple already reviewed Pull Requests.
3. Only consists of merge commits.

Since each pull request has already been reviewed before being merged into the feature branch, the
feature branch review is more of a sanity check. The reviewer should ensure the following:

- The feature branch is ready to be merged into master.
  - Ensure the work has been QA tested, including writing QA notes should it be needed.
  - Or verify the feature is behind a feature flag, and the crossover boundaries goes through
    testing.
- Review any non-reviewed commits made directly on the branch, these can be either feature work or
  merge commits. For more details see [Reviewing](#reviewing).

:::warning

Since feature branches do not have the same protections as master, it's technically possible to
commit directly to the branch or merge a pull request without a up-to-date review. However this
should be discouraged, and should be avoided whenever possible. The only exception being merge
commits.

:::

## Syncing Feature Branch

We typically denote one person as being responsible for keeping the feature branch up to

> date.<bitwarden> In most cases this will be the TPC of the pod but can be anyone.</bitwarden> This

Reviewer 3: ```suggestion
date.<bitwarden> In most cases this will be the Tech Lead of the pod but can be anyone.</bitwarden> This
```

Author: TPC is not tech lead though.

---
sidebar_custom_props:
  access: bitwarden
---

# Long-lived Feature Branches

A long-lived feature branch differs from a regular Pull Requests in three significant ways. It
should:

- Have multiple authors

> - Consists of multiple already reviewed Pull Requests

Reviewer: ```suggestion
- Consist of multiple Pull Requests, which have already been reviewed
```

> - Only consists of merge commits

Reviewer: ```suggestion
- Only consist of merge commits
```
